### PR TITLE
Add tests for SMAC

### PR DESCRIPTION
**Pull request by y0z**
## Contributor Agreements

Please read the [contributor agreements](https://github.com/optuna/optunahub-registry/blob/main/CONTRIBUTING.md#contributor-agreements) and if you agree, please click the checkbox below.

- [x] I agree to the contributor agreements.

## Motivation

Add tests for SMAC to ensure its quality.

## Description of the changes

- Add tests for SMAC based on the features currently supported.

## Notes

Below is a summary of the features missing from `SMACSampler` that I noticed while introducing the tests (and I skipped the tests for such features at the moment).

- Support of [`reseed_rng`](https://optuna.readthedocs.io/en/stable/reference/samplers/generated/optuna.samplers.BaseSampler.html#optuna.samplers.BaseSampler.reseed_rng) (importance: relatively high)
  - If this feature is missing, parallelization using `n_jobs` will not work as expected since `study.optimize` internally calls this method (cf.  https://github.com/optuna/optuna/blob/467cca671e4940145f2feffe173dfcfc153b5744/optuna/study/_optimize.py#L102-L138). However, supporting this method seems non-trivial since SMAC does not support reseeding. As a temporary workaround, I added a warning to inform users that `reseed_rng` does not work.
- Support of the following types of search space (importance: low)
  - conditional search space
  - dynamic distributions (i.e., changing its lower/upper bound during optimization)
  - search space with a single value (i.e., low == high holds)
- Handling of the NaN objective value (importance: low)

### CODE REVIEW

File: package/samplers/smac_sampler/sampler.py [hunk 1]
```diff
@@ -344,5 +353,5 @@ def _convert_to_config_space_design_space(
             else:
                 raise NotImplementedError(f"Unknown Hyperparameter Type: {type(distribution)}")
             if hp is not None:
-                config_space.add_hyperparameter(hp)
+                config_space.add(hp)
```
Comment (y0z) [inline package/samplers/smac_sampler/sampler.py:356]: `config_space.add_hyperparameter` is deprecated and it is recommended to use `add`.

File: package/samplers/smac_sampler/sampler.py [hunk 2]
```diff
@@ -237,6 +243,9 @@ def sample_relative(
         study._storage.set_trial_system_attr(trial._trial_id, _SMAC_SEED_KEY, trial_info.seed)
         params = {}
         for name, hp_value in cfg.items():
+            # SMAC uses np.int64 for integer parameters
+            if isinstance(hp_value, np.int64):
+                hp_value = hp_value.item()  # Convert to Python int.
```
Comment (y0z) [inline package/samplers/smac_sampler/sampler.py:248]: SMAC returns `np.int64`, and it fails some tests. This change fixes it.

File: package/samplers/smac_sampler/sampler.py [hunk 3]
```diff
@@ -226,6 +227,11 @@ def _get_init_design(
             raise NotImplementedError(f"Unknown Initial Design Type: {init_design_type}")
         return init_design
 
+    def reseed_rng(self) -> None:
```
Comment (y0z) [inline package/samplers/smac_sampler/sampler.py:230]: `reseed_rng` is currently not implemented. This will cause unexpected behavior when `n_jobs > 1`.

**Comment by nabenabe0928** [comment]
@dengdifan 
Hey, are there any ways to update the seed of SMAC after we start optimization?
We need it to increase the randomness of sampler when SMAC is used in multi-threading.

**Comment by HideakiImamura** [comment]
@nabenabe0928 Could you review this PR?

**Review by nabenabe0928 — COMMENTED** [review]
Thank you for the PR, I left a comment below!

Why are they not tested?
- test_sample_relative_categorical
- test_sample_single_distribution
- test_single_parameter_objective
- test_combination_of_different_distributions_objective

Could you add comment lines for the omits?
- test_reproducible_in_other_process
- test_reproducible
- test_sampler_reseed_rng

Can we maybe raise an error or warn in this scenario instead of skipping them?
- test_nan_objective_value
- test_dynamic_range_objective
- test_conditional_parameter_objective

File: package/samplers/smac_sampler/tests/test_sampler.py [hunk 4]
```diff
@@ -0,0 +1,498 @@
...
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT
+NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
+DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT
+OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+This file is taken from Optuna (https://github.com/optuna/optuna/blob/master/tests/samplers_tests/test_samplers.py)
+and modified to test SMACSampler.
+"""
```
Comment (nabenabe0928) [inline package/samplers/smac_sampler/tests/test_sampler.py:22]:
```suggestion

Note that the following tests were omitted for now:
- test_reproducible_in_other_process
- test_reproducible
- test_nan_objective_value
- test_dynamic_range_objective
- test_conditional_parameter_objective

"""
```

**Review by nabenabe0928 — APPROVED** [review]

**Pull request merged**